source/common/upstream/ring_hash_lb.h

```diff
@@ -39,22 +41,48 @@ class RingHashLoadBalancer : public LoadBalancerBase,
   };
 
   struct Ring {
-    HostConstSharedPtr chooseHost(LoadBalancerContext* context, Runtime::RandomGenerator& random);
-    void create(const Optional<envoy::api::v2::Cluster::RingHashLbConfig>& config,
-                const std::vector<HostSharedPtr>& hosts);
+    Ring(const Optional<envoy::api::v2::Cluster::RingHashLbConfig>& config,
+         const std::vector<HostSharedPtr>& hosts);
+    HostConstSharedPtr chooseHost(LoadBalancerContext* context,
+                                  Runtime::RandomGenerator& random) const;
 
     std::vector<RingEntry> ring_;
   };
 
-  void refresh(uint32_t priority);
+  typedef std::shared_ptr<const Ring> RingConstSharedPtr;
 
-  const Optional<envoy::api::v2::Cluster::RingHashLbConfig>& config_;
-  struct PerPriorityState {
-    Ring all_hosts_ring_;
-    Ring healthy_hosts_ring_;
+  struct LoadBalancerImpl : public LoadBalancer {
+    LoadBalancerImpl(ClusterStats& stats, Runtime::RandomGenerator& random,
+                     const RingConstSharedPtr& ring, bool global_panic)
+        : stats_(stats), random_(random), ring_(ring), global_panic_(global_panic) {}
+
+    // Upstream::LoadBalancer
+    HostConstSharedPtr chooseHost(LoadBalancerContext* context) override;
+
+    ClusterStats& stats_;
+    Runtime::RandomGenerator& random_;
+    const RingConstSharedPtr ring_;
+    const bool global_panic_;
+  };
+
+  struct LoadBalancerFactoryImpl : public LoadBalancerFactory {
+    LoadBalancerFactoryImpl(ClusterStats& stats, Runtime::RandomGenerator& random)
+        : stats_(stats), random_(random) {}
+
+    // Upstream::LoadBalancerFactory
+    LoadBalancerPtr create() override;
+
+    ClusterStats& stats_;
+    Runtime::RandomGenerator& random_;
+    std::shared_timed_mutex mutex_;
+    RingConstSharedPtr current_ring_;
```

Comment:
Did we ever decide if we wanted to use https://clang.llvm.org/docs/ThreadSafetyAnalysis.html annotations for mutex protected variables?

Reply:
Sure let me figure that out.

Reply:
```
In file included from source/common/upstream/ring_hash_lb.cc:1:
bazel-out/darwin-fastbuild/bin/source/common/upstream/_virtual_includes/ring_hash_lb_lib/common/upstream/ring_hash_lb.h:80:38: error: 'guarded_by' attribute requires arguments whose type is annotated with 'capability' attribute; type here is 'std::shared_timed_mutex' [-Werror,-Wthread-safety-attributes]
    RingConstSharedPtr current_ring_ GUARDED_BY(mutex_);
                                     ^
external/com_google_absl/absl/base/thread_annotations.h:54:53: note: expanded from macro 'GUARDED_BY'
#define GUARDED_BY(x) THREAD_ANNOTATION_ATTRIBUTE__(guarded_by(x))
                                                    ^
In file included from source/common/upstream/ring_hash_lb.cc:1:
bazel-out/darwin-fastbuild/bin/source/common/upstream/_virtual_includes/ring_hash_lb_lib/common/upstream/ring_hash_lb.h:81:24: error: 'guarded_by' attribute requires arguments whose type is annotated with 'capability' attribute; type here is 'std::shared_timed_mutex' [-Werror,-Wthread-safety-attributes]
    bool global_panic_ GUARDED_BY(mutex_){};
                       ^
external/com_google_absl/absl/base/thread_annotations.h:54:53: note: expanded from macro 'GUARDED_BY'
#define GUARDED_BY(x) THREAD_ANNOTATION_ATTRIBUTE__(guarded_by(x))
                                                    ^
source/common/upstream/ring_hash_lb.cc:51:19: error: reading variable 'current_ring_' requires holding mutex 'mutex_' [-Werror,-Wthread-safety-analysis]
    ring_to_use = current_ring_;
                  ^
source/common/upstream/ring_hash_lb.cc:52:27: error: reading variable 'global_panic_' requires holding mutex 'mutex_' [-Werror,-Wthread-safety-analysis]
    global_panic_to_use = global_panic_;
                          ^
source/common/upstream/ring_hash_lb.cc:175:13: error: writing variable 'current_ring_' requires holding mutex 'factory_->mutex_' exclusively [-Werror,-Wthread-safety-analysis]
  factory_->current_ring_ = new_ring;
            ^
source/common/upstream/ring_hash_lb.cc:176:13: error: writing variable 'global_panic_' requires holding mutex 'factory_->mutex_' exclusively [-Werror,-Wthread-safety-analysis]
  factory_->global_panic_ = new_global_panic;```

OSX clang seems to not be happy w/ the annotations around shared mutex. In the interest of time I'm going to remove them and add a TODO.